kvserver: don't quiesce when there's outstanding quota

andreimatei opened the pull request:
See individual commits.

andreimatei committed:
kvserver: don't quiesce when there's outstanding quota

In #46699 we've seen that quiescing a range while there's outstanding
replication quota can lead the range deadlock, since there might be no
quota available to propose a command that would un-quiesce the range.

This patch makes it so that we don't quiesce with outstanding quota.

Fixes #46699

Release note (bug fix): A rare bug causing a range to deadlock and all
the writes to the respective range to timeout was fixed.

andreimatei committed:
kvserver: fix updating of liveness map

The store caches the results of NodeLiveness.GetIsLiveMap(). The cache
was updated in two places. One of the two places was performing
massaging of the liveness results, the other one wasn't. This looks like
a mistake; now they both do the massaging.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48082)
<!-- Reviewable:end -->

andreimatei commented:
My brothers, please look over this strawman. The jungle is deep and fragile and I'd like y'all's opinions on the record.

This PR is about rationalizing as much as possible the relationship between quiescing a range and the maintenance of its proposal quota. I'm struggling on two axes:
1) What should be the relationship, if any, between the decision to quiesce (which is currently mainly done based on the Raft status of all the followers plus some liveness) and the decision to release quota held by outstanding commands (which is based also on the Raft status, but more indirectly, plus some liveness).
2) What liveness signals should we use for the quiescing and quota release decisions and, based on those signals, what exactly should the policy be? When should we give up on behind followers for one and/or for the other?

Hopefully this will make sense after looking at the last commit.

tbg reviewed (commented):
Thanks for working on this! It'll be nice to get this old code polished up a notch.

Turns out that... I'm doubtful that it's a good goal to use the same condition in quota release and quiescing:

The network partition thing is a real threat, but the more likely threat is on the quota release side. If we're lenient in considering followers live, we're setting us up for scenarios in which we run out of quota. We really just don't want the same criterion here:

In quiescing, we want to be lenient about considering followers active. If there's a chance they're around, consider them and don't quiesce. This is because the prime danger is quiescing tons of ranges based on some system failure, in which case these ranges won't eagerly repair themselves until next touched which is bad. Not quiescing tons of ranges at the same time is also bad, but you have to expect that to happen when you set up your system anyway. Relying on most ranges being quiesced is just not something that will be possible during failures in general.

In quota release, we want to be strict about considering followers active. If there's reasonable doubt that they may not be able to make progress, ignore them. This is because failing to recognize an inactive follower as such will make ranges unavailable as they run out of quota. I suspect that this is the biggest enemy here and this PR sets us up to hit this more frequently: any disk stall will do it now, while before this PR, I think, we'd ignore such a node because its lastUpdateTimes would be stale, or at least there's some hope because this is only updated when we step a message into the leader (and the raft loop on the follower with a stuck disk would be stuck):

https://github.com/cockroachdb/cockroach/blob/a72b871dfb16322eb54497ab99c03107e11dc36a/pkg/kv/kvserver/replica_raft.go#L361

All of this suggests that we shouldn't try to make quota and quiescing use the exact same criterion. Instead, there are two flavors:

- is follower likely to be able to make progress? Used by quota pool -> last update times, and a more lenient version:
- is follower possibly active? Used by quiescence --> likely to make progress || RPC connection

Adding the liveness record into the mix is possible, but I don't know if I find it useful given the increase in complexity. If we add it to the "make progress" condition, we're just saying "I don't trust the last update times". (Or perhaps there's a reason in a perfectly healthy system for the last update times to be out of date). If we add it to the quiescence condition, we have what's done in this PR, where a live node gets a higher timeout. That's fine by me too, though I personally would avoid the complexity because the only upshot is that we would sometimes quiesce during a partition, but that seems lower on the list of problems.

----

> A node is live if either we have a healthy gRPC conn to it, or
lastUpdateTime is not stale. Either of these are signals that the range
can make progress.

I think it's worth putting into writing again the argument from https://github.com/cockroachdb/cockroach/pull/27617: quiescing can't be too aggressive or a broken cluster will aggressively quiesce every range.

Also, "A replica is live", not a "a node is live".

I'm seeing some ambiguity about the use of the words optimism/pessimism here. Quiescing optimistically would mean to me that when in doubt, you quiesce, meaning that we'd mark nodes as dead aggressively. You write

> On the optimistic/pessimistic front, we're now optimistic for both quota
and quiescing: we don't quiesce and we don't release quota while there's
ambiguity about the liveness of a follower that's behind.

which means that you've got it the other way around. I think this is because you are talking about the concept of availability, I think you want to clarify that throughout (there's one more mention in the PR message).

Also, should we avoid the term "liveness" here in the first place? What's happening here is distinctly not node liveness and yet there's a livenessChecker, etc. All of this won't add clarity. This is also not a natural replacement for node liveness. For example, a node with stalled disks won't mark itself as live any more (in the node liveness sense) but we'd be pretty sad if the replicate queue (using the livenessChecker) would decide that all is well and that replica can just stay there. It's "fine" (given the alternatives) to not quiesce ranges with such a replica, but such a replica is decidedly "not live". Similarly, such a node will hog quota (probably same before this change, though) and stall the ranges it has. We may want to call a spade a spade and stick with the `active` or `online` terminology: we're really mostly checking that a follower is there. We're not checking that it can make progress.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-:-M6-LNT18E2e4KooDTBJ:b7rxvp2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @bdarnell, @nvanbenschoten, and @petermattis)

___
*[pkg/kv/kvserver/replica_raft.go, line 339 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-Fasm0EWOYuyoGSVE:-M6-Fasn5j59ag4DgBSR:b-5pcm0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2241b62bb1cb94869eac1da24797bc068908dc00/pkg/kv/kvserver/replica_raft.go#L339)):*
> ```Go
> 
> func (r *Replica) numPendingProposalsRLocked() int {
> 	return len(r.mu.proposals) + r.mu.proposalBuf.Len()
> ```

Looks like there's something to do here. The only other caller is the range state, so maybe it's fine? But everyone new should really use `hasPendingProposals`.
___
*[pkg/kv/kvserver/replica_raft.go, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-FjZz9EKRjcSX6LhH:-M6-FjZz9EKRjcSX6LhI:bse59ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/2241b62bb1cb94869eac1da24797bc068908dc00/pkg/kv/kvserver/replica_raft.go#L346)):*
> ```Go
> // has been proposed by the current replica and it hasn't been applied locally
> // yet. A command also counts as outstanding if it's holding up quota - i.e. if
> // it hasn't been persisted in all the other live replicas' logs. This second
> ```

nit: if it hasn't been acked by all other live replicas.
___
*[pkg/kv/kvserver/replica_raft.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-H4oa9jPAWVyXydQR:-M6-H4oa9jPAWVyXydQS:b56ezqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2241b62bb1cb94869eac1da24797bc068908dc00/pkg/kv/kvserver/replica_raft.go#L351)):*
> ```Go
> // quiesced, and it might prevent the range from unquiescing (leading to
> // deadlock).
> func (r *Replica) hasPendingProposalsRLocked() bool {
> ```

Should this be called `hasAllFollowersCaughtUp` or something like that? The current name is now misleading.
___
*[pkg/kv/kvserver/replica_raft.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-NMKU3V4lhJiJIvcb:-M6-NMKU3V4lhJiJIvcc:bayakjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft.go#L350)):*
> ```Go
> }
> 
> // hasPendingProposalsRLocked returns true if this range cannot be quiesced
> ```

name stale
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 170 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-NWvv0uhEFX08L9Px:-M6-NWvv0uhEFX08L9Py:b-pkbykl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L170)):*
> ```Go
> 	checker livenessChecker, rep roachpb.ReplicaDescriptor, now time.Time,
> ) bool {
> 	if checker == nil {
> ```

This is probably for unit tests, but can you check whether this is easy to avoid by fixing up the tests?
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 216 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-NoQb-cskdMQ7_MwB:-M6-NoQb-cskdMQ7_MwC:bl9hgk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L216)):*
> ```Go
> // and also we don't want to quiesce if another replica might have commands
> // pending that require this leader for proposing them.
> // b) There's no commands in-flight proposed by this leaseholder. Clients can be
> ```

There are
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 219 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-Nsnj9GhQnJ3r7QQt:-M6-Nsnj9GhQnJ3r7QQu:b458qhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L219)):*
> ```Go
> // waiting for results while there's pending proposals.
> // c) There's no outstanding quota. Quiescing while there's quota outstanding
> // can lead to deadlock.
> ```

refer to an issue
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-O5o08EeY_yvyZD1J:-M6-O5o08EeY_yvyZD1K:b-8dqh49) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L243)):*
> ```Go
> 	// determination as we do below, so it's weird that we allow the
> 	// hasPendingQuotaRLocked() info to be "stale".
> 	// An alternative is to not look at the pending quota at all, and instead just
> ```

I'm in favor of clearing it when we do quiesce. In a system where most ranges are quiesced, this translates to a bit of saved memory, while at the same time avoiding state changes (mod the clearing) of the quota pool triggered from one-off paths (this one).


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> In quiescing, we want to be lenient about considering followers active. If there's a chance they're around, consider them and don't quiesce. This is because the prime danger is quiescing tons of ranges based on some system failure, in which case these ranges won't eagerly repair themselves until next touched which is bad.

Keep in mind that we'll also un-quiesce as soon as a replica's liveness record goes from dead to alive.

> In quota release, we want to be strict about considering followers active. If there's reasonable doubt that they may not be able to make progress, ignore them.

To ponder things more, let me ask you this: do you like the `lastUpdateTimes` map, as a per-range structure? Cause you have a TODO on it to get rid of it and centralize things on the liveness record, or something like it.
My PR was kinda taking steps compatible with eventually deleting this structure - the `lastUpdateTimes` was not used directly any more.
I find the concept of per-range progress dubious. If a node is able to make progress on some range, it should be able to make progress on any range. If the node is not able to make progress on any range, then its liveness record should also reflect this somehow. A "disk stall" should cause leases to go away, if it's long enough. It should also cause a decommission I think, if it's even longer. What is the degradation mode where we want the quota pool to not track the node, but we otherwise want the things that depend on the liveness record (leases) to continue?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-:-M6041oG8RGhycfhukf3:b-frjtau)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @bdarnell, @nvanbenschoten, and @petermattis)



<!-- Sent from Reviewable.io -->

bdarnell commented:
> Turns out that... I'm doubtful that it's a good goal to use the same condition in quota release and quiescing:

I've been advocating using the same signal here, but I think I'm persuaded that it's good that they're different. We just need to document *why* they're different, and why they're using signals that don't even appear to be directly comparable to each other. And ideally I think I'd like to consolidate all our different liveness signals for easy comparison and so that people can reuse one that fits instead of inventing yet another ad-hoc mechanism. 

> I find the concept of per-range progress dubious. If a node is able to make progress on some range, it should be able to make progress on any range.

In general, I'd expect us to evolve towards more fine-grained notions of health and isolation, so a node may be able to make progress on some ranges but not others (consider non-filesystem-based cloud storage). That's relatively distant, though, so a per-range progress map may be premature at this point.

andreimatei committed:
kvserver: assorted quiescing comments

Release note: None

andreimatei committed:
kvserver: homogenize liveness threshold for followers

... across quota releasing and quiescing.

Quiescing is supposed to have a higher bar for giving up on waiting for
a node because the node is dead than releasing proposal quota. However,
the latter was using a higher inactivity time threshold (10s, versus the
liveness record duration in the case of quiescing - which defaults to
9s). This patch makes quota releasing use the same 9s.
Liveness determinations across the two processes continue to not be
synchronized, and the threshold is applied to different activity
signals, but still - making them uniform supports the story that the
quota release condition is supposed to be morally a superset of the
quiescence once.

Release note: None

andreimatei reviewed (commented):
I've reworked this patch-set to fix the bug while being minimally invasive, cause I'm getting discouraged. I don't really know what the best thing to do is. PTAL.

The patch no longer touches the different policies. It tries to explain them a little bit. And it fiddles slightly with a threshold.

> I've been advocating using the same signal here, but I think I'm persuaded that it's good that they're different. We just need to document why they're different, and why they're using signals that don't even appear to be directly comparable to each other.

For the sake of the conversation, it's worth noting that, while the intention was for the quota releasing liveness determination to be more strict than the quiescing one, it was using a *higher* inactivity threshold - 10s vs 9s. I've now made them use the same threshold, fwiw, even though the threshold continues to apply to different things.

@tbg, if you care, please spell out exactly what you think the signals that go into the two policies should be - even more than you've spelled them out last time. I'll also take a "let's move on with our lives" now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-:-M61wNwR-AZf8le-Ct7O:b-ey2jrb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bdarnell, @nvanbenschoten, @petermattis, and @tbg)

___
*[pkg/kv/kvserver/replica_raft.go, line 339 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-Fasm0EWOYuyoGSVE:-M61VHFg1CDPJ4gAf4pD:b2h3b1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2241b62bb1cb94869eac1da24797bc068908dc00/pkg/kv/kvserver/replica_raft.go#L339)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Looks like there's something to do here. The only other caller is the range state, so maybe it's fine? But everyone new should really use `hasPendingProposals`.
</blockquote></details>

So what exactly are you suggesting? The method is not part of the `quiescer` interface, if that's what you were driving at.

I don't have a particular problem with it...
___
*[pkg/kv/kvserver/replica_raft.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-H4oa9jPAWVyXydQR:-M61ViTj5aVFnjPVQ6G-:b-2x4j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/2241b62bb1cb94869eac1da24797bc068908dc00/pkg/kv/kvserver/replica_raft.go#L351)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Should this be called `hasAllFollowersCaughtUp` or something like that? The current name is now misleading.
</blockquote></details>

see now
___
*[pkg/kv/kvserver/replica_raft.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-NMKU3V4lhJiJIvcb:-M61WjMmEn5L3T1wL2p-:b2ni76j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft.go#L350)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

name stale
</blockquote></details>

n/a
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 170 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-NWvv0uhEFX08L9Px:-M61wFczE0N5WCxkeVUq:b2ni76j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L170)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is probably for unit tests, but can you check whether this is easy to avoid by fixing up the tests?
</blockquote></details>

n/a
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 216 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-NoQb-cskdMQ7_MwB:-M61sG4m1bvheW9-e_Da:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L216)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

There are
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 219 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-Nsnj9GhQnJ3r7QQt:-M61wD6PCwjDALzY5Uqy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L219)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

refer to an issue
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-O5o08EeY_yvyZD1J:-M61wBqz1RFS6etbUgtj:b2ni76j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L243)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm in favor of clearing it when we do quiesce. In a system where most ranges are quiesced, this translates to a bit of saved memory, while at the same time avoiding state changes (mod the clearing) of the quota pool triggered from one-off paths (this one).
</blockquote></details>

n/a


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48082)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bdarnell, @nvanbenschoten, @petermattis, and @tbg)

___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-M6-O5o08EeY_yvyZD1J:-M621Ah75xlB-lB-lO0W:b-3gwum8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e79e437d0f53ebf914f2acf9b76001d73e47e7d/pkg/kv/kvserver/replica_raft_quiesce.go#L243)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

n/a
</blockquote></details>

oh btw @nvanbenschoten and @ajwerner spoke against this option cause it'd be one more place where we muck with the quota. Although it always sounded good to me...


<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1523,7 +1530,7 @@ func (m lastUpdateTimesMap) isFollowerActive(
 		// replicas were updated).
 		return false
 	}
-	return now.Sub(lastUpdateTime) <= MaxQuotaReplicaLivenessDuration
```
Is this now unused? Looks like it to me. Some tests use it but those need fixing?

andreimatei replied:
It still had a user outside of tests. But I've added a commit to get rid of it. PTAL.

tbg commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1511,10 +1511,17 @@ func (m lastUpdateTimesMap) updateOnBecomeLeader(descs []roachpb.ReplicaDescript
 
-// isFollowerActive returns whether the specified follower has made
-// communication with the leader in the last MaxQuotaReplicaLivenessDuration.
+// isFollowerActive is like isFollowerActiveSince, with a default threshold.
```
I think it might clarify things a little bit if this always required the parameter to be passed in.

andreimatei replied:
done

tbg reviewed (approved):
Code changes LGTM. Add a unit test for the bug fix though, please.

andreimatei committed:
kvserver: remove lastUpdateTimes.isFollowerActive() and MaxQuotaReplicaLivenessDuration

... in favor of lastUpdateTimes.isFollowerActiveSince(). Make all
callers be explicit about what inactivity threshold they want - and make
all of them use the liveness record duration.

Release note: None

andreimatei commented:
> Add a unit test for the bug fix though, please.

You mean for the "fix updating of liveness map" commit? I wouldn't know how to add that test, to be honest. There's no test for `nodeIsLiveCallback()` (in fact everything passes if I comment it out :) ). And even if there was a test for that method, I'd still be unsure about how to test the change... I'll leave it for next time.

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-:-M6AYfti4cmV5cyXUz6N:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bdarnell, @nvanbenschoten, @petermattis, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
The failure was an unrelated timeout.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48082#-:-M6BCyQICTSWMdchOuuK:btfdcxq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bdarnell, @nvanbenschoten, @petermattis, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei commented:
bors r+
